### ISSUE TITLE
Error when visualizing with GLVis the solutions of ex15

### ISSUE THREAD

**Issue by psocratis** [mfem/mfem#1843]
Dear all, 

There seems to be an issue when trying to visualize consecutive solutions in GLVis produced by `ex15/ex15p`. 
A minimal example run that causes the problem is `./ex15`, and it appears at `Time = 0.23` with 
`Error in send(): Broken pipe`

The error appears in both serial and parallel builds of mfem (`master` branch) in debug mode. This also happens with both glvis-3.4 and the latest `modern-ogl-dev` branch of GLVis. 

My compiler/OS configuration is `Apple clang version 12.0.0 (clang-1200.0.32.21)` on `MacOS Catalina Version 10.15.7(19H2)`.

Thanks, 

Socratis

**Comment by termi-official** [comment]
Hi,
I also ran into the same issue, but failed to track down find what exactly causes it. See https://github.com/GLVis/glvis/issues/105 . My workaround was to simply feed data less frequently to GLVis.
Best,
Dennis

**Comment by pazner** [comment]
This is caused by the Mac C++ stream library setting the failbit when parsing very small doubles (smaller than the smallest normal double, i.e. about 2.2251e-308). This is illustrated with the following short example:
```c++
#include <iostream>
#include <ios>
#include <sstream>

int main()
{
   std::stringstream str;
   double x;
   str << "3e-308\n";
   str >> x;
   std::cout << std::scientific << "x = " << x << '\n';
   if (!str) std::cerr << "Error:   " << std::strerror(errno) << std::endl;
   
   str << "2e-308\n";
   str >> x;
   std::cout << std::scientific << "x = " << x << '\n';
   if (!str) std::cerr << "Error:   " << std::strerror(errno) << std::endl;
   
   return 0;
}
```
that on Mac should give the output
```
x = 3.000000e-308
x = 2.000000e-308
Error:   Result too large
```

There are no issues parsing 3e-308, but 2e-308 gives an error. Notice that the number 2e-308 is actually parsed correctly, but the failbit on the stream is still set.

We can work around this issue by modifying `Vector::Load` in MFEM:
```c++
void Vector::Load(std::istream &in, int Size)
{
   SetSize(Size);

   for (int i = 0; i < size; i++)
   {
      in >> data[i];
#if defined __APPLE__
      if (!in && errno == ERANGE)
      {
         in.clear();
      }
#endif
   }
}
```

For me that resolves the issue with GLVis and example 15 (after rebuilding GLVis).

**Comment by v-dobrev** [comment]
Interesting ... I though the issue is on the sending side of the socket. How did you find this? It seems very tricky to find. 😄

**Comment by tzanio** [comment]
Amazing detective work @pazner !!

**Comment by psocratis** [comment]
Wow that's awesome, I'll give it a try and report back

**Comment by pazner** [comment]
> Interesting ... I though the issue is on the sending side of the socket. How did you find this? It seems very tricky to find. 😄

I found it through judicious use of printf debugging, with the benefit of having seen similar parsing errors occur in the past. This same problem causes issues in ParaView on Mac when loading ASCII VTK files with subnormal numbers.

**Comment by psocratis** [comment]
Works for me too!

**Comment by tzanio** [comment]
I can confirm that this fixes the issue for me too 😮 

@pazner do you want to add the fix in the `mfem-4.2-dev` branch?

**Comment by v-dobrev** [comment]
Is this an issue with gcc on Mac? Maybe the `#ifdef` should be just for Apple clang?

**Comment by pazner** [comment]
My guess would be that this is not an issue with GCC, but I haven't checked. (I believe it is a feature/bug of clang's libc++)

**Comment by tzanio** [comment]
Resolved by #1847. Thanks @pazner!
